stdlib/public/Concurrency/Task.cpp

```diff
@@ -21,20 +21,129 @@
+  assert(isFuture());
+  auto fragment = futureFragment();
+
+  auto currentStatus = fragment->status.load();
+  switch (currentStatus) {
+  case FutureFragment::Status::Error:
+  case FutureFragment::Status::Success:
+    // The task is done; we don't need to wait.
+    return currentStatus;
+
+  case FutureFragment::Status::Executing:
+    break;
+  }
+
+  // Put the waiting task at the beginning of the wait queue.
+  while (true) {
+    waitingTask->NextWaitingTask = fragment->waitQueue.load();
+    if (fragment->waitQueue.compare_exchange_strong(
+            waitingTask->NextWaitingTask, waitingTask)) {
+      // Escalate the priority of this task based on the priority
+      // of the waiting task.
+      swift_task_escalate(this, waitingTask->Flags.getPriority());
+      break;
+    }
+  }
+
+  return FutureFragment::Status::Executing;
+}
+
+void AsyncTask::completeFuture(AsyncContext *context) {
+  assert(isFuture());
+  auto fragment = futureFragment();
+  auto storagePtr = fragment->getStoragePtr();
+
+  // Check for an error.
+  if (unsigned errorOffset = fragment->errorOffset) {
+    // Find the error object in the context.
+    auto errorPtrPtr = reinterpret_cast<char *>(context) + errorOffset;
+    OpaqueValue *errorObject = *reinterpret_cast<OpaqueValue **>(errorPtrPtr);
+
+    // If there is an error, take it and we're done.
+    if (errorObject) {
+      *reinterpret_cast<OpaqueValue **>(storagePtr) = errorObject;
+      fragment->status = FutureFragment::Status::Error;
+      return;
+    }
+  }
+
+  // Take the success value.
+  auto resultPtr = reinterpret_cast<OpaqueValue *>(
+      reinterpret_cast<char *>(context) + fragment->resultOffset);
+  fragment->resultType->vw_initializeWithTake(storagePtr, resultPtr);
+  fragment->status = FutureFragment::Status::Success;
+}
+
+void AsyncTask::scheduleWaitingTasks(ExecutorRef executor) {
+  assert(isFuture());
+  auto fragment = futureFragment();
+
+  auto waitingTask = fragment->waitQueue.load();
+
+  // Unnecessary, but useful for the assertion at the end.
+  fragment->waitQueue = nullptr;
+
+  while (waitingTask) {
+    // Find the next waiting task.
+    auto nextWaitingTask = waitingTask->NextWaitingTask;
+
+    // Remove this task from the list.
+    waitingTask->NextWaitingTask = nullptr;
+
+    // TODO: schedule this task on the executor.
+
+    // Move to the next task.
+    waitingTask = nextWaitingTask;
+  }
+
+  // Nobody should be able to add anything to the queue in this function.
+  assert(!fragment->waitQueue.load());
+}
 
 SWIFT_CC(swift)
-static void destroySimpleTask(SWIFT_CONTEXT HeapObject *_obj) {
-  auto obj = static_cast<AsyncTask*>(_obj);
+static void destroySimpleTask(SWIFT_CONTEXT HeapObject *obj) {
+  auto task = static_cast<AsyncTask*>(obj);
+
+  // FIXME: "Simple task" seems like it's not that useful a concept. We might
+  // also have task-local storage to destroy.
+  // assert(!task->isFuture());
```

Comment:
I agree that it probably makes sense to just have the one task-destruction function.  You can just rename this.